Test refactoring: new test-suite via pytest

Zempashi opened the pull request:
Fixes: #116

Commit:
Test documentation

Commit:
Refactoring of test-suites using pytest and pytest-bdd

Zempashi commented on tox.ini:
```diff
@@ -34,3 +34,15 @@ commands =
+skip_install = true
+deps =
+    -r{toxinidir}/tests/requirements.txt
+passenv = *
```
Required to be able to use ssh-agent

NicolasT replied:
Then rather pass specific variable names, and a comment to explain why.

Zempashi commented on tests/test_installation.py:
```diff
@@ -0,0 +1,27 @@
+def inventory_bdd(inventory):
+    return inventory
+
+@when("I run 'make shell'")
```
Do we want to sandbox tests (avoid cache hit on a local dev laptop). Avoiding cache hit will slow down test, and make internet sensitive (flakiness ?)

NicolasT replied:
Nah, relying on an existing `make shell` environment should be fine.

NicolasT commented on tests/README.md:
```diff
@@ -0,0 +1,90 @@
+MetalK8s is an infrastructure product, that rely on ansible so framework in python is somewhat “obvious”.
+Pytest is quite famous right famous.
```
I'll only look at the content for now, not language :wink:

Zempashi replied:
"right now". yeah bear with me

NicolasT commented on tests/README.md:
```diff
@@ -0,0 +1,90 @@
+
+1. Test that run before installation on any machine
+    - test of os-check (centos check and kernel)
+    - test first installation of component that doesn't need kubernetes (node_exporter)
```
So, test roles in isolation.

Zempashi replied:
But could be more than one role, could also be "storage"

NicolasT replied:
Shouldn't that ideally be provisioned by one role?

Zempashi replied:
Maybe I over developed my ansible "theory",but: no. A role only target one "set" of server. Meaning that we want a set of feature only be developed using one set of server is too restrictive for me (example storage: create storage on each node, but only register on master node)

NicolasT replied:
Ah, I see. Fair enough, though most of the time this will involve interaction with a K8s cluster, which will be the 'difficult' roles to test anyway ;)

NicolasT commented on tests/README.md:
```diff
@@ -0,0 +1,90 @@
+    - test first installation of component that doesn't need kubernetes (node_exporter)
+    - test of the complete installation it-self
+
+2. Test that run after the installation of kubernetes only (not the entire set of MetalK8s):
```
I don't understand this.

Zempashi replied:
For exemple role kube_prometheus install prometheus via helm, when you don't want to test an upgrade, you need test in "isolation" like you already mentioned. But for kube_prometheus we require kubernetes to be up. May be we could say that we can't test the  `.2`

NicolasT replied:
Yeah, roles that interact with K8s are... tricky :-/ Doesn't fit the Ansible model all too well anyway.

NicolasT commented on tests/README.md:
```diff
@@ -0,0 +1,90 @@
+4. Test that meant to be run after installation and before production (conformance test)
+    - test kubernetes object creation
+
+5. Test that need to run on non-production environnement (CI test)
```
Include K8s Conformance suite.

Zempashi replied:
Indeed

Zempashi replied:
Considering the time to run conformance, we might not be able to run it at each commit, or merge. May be that's something we can have at each "release" (including alpha or rc to see problem before needing the release). IHM, we do not need to integrate this in the pytest framework (and it takes 2 hours, no parallelism possible), but keeping the sonobuoy runner that we could run directly from our CI

NicolasT replied:
Fair enough. Kind of a 'nightly' (and indeed, pre-release).

NicolasT commented on tests/README.md:
```diff
@@ -0,0 +1,90 @@
+6. Test that could run with only mocked environement:
+    - inventory check
+
+7. Unittesting of ansible roles: molecule
```
How does this relate to item 1?

Zempashi replied:
IHM, no relation.

NicolasT commented on tests/README.md:
````diff
@@ -0,0 +1,90 @@
+7. Unittesting of ansible roles: molecule
+
+```
+pytest --inventory <path_to_inventory> --install --conformance --go-destructive
````
I'd rather use test selectors than custom command line arguments.

Zempashi replied:
ok no problem, doesn't realy change much the "design"

NicolasT commented on tests/README.md:
```diff
@@ -0,0 +1,90 @@
+
+* `--conformance` enable conformance test (handles `.4`)
+
+* `--go-destructive` enable full testing, including test of failover (handles `.5`)
```
`go`?

NicolasT commented on tests/README.md:
```diff
@@ -0,0 +1,90 @@
+No args on pytest mean only non-destructive test (handle case `.3` and `.6`)
+
+This represents e2e/integration testing of MetalK8s. Might be better to
+have unittesting as a separate test-suite (Molecule, different python scripts in the repo)
```
Yeah, I looked at how to 'integrate' Molecule tests in pytest, but didn't find a readily-available solution. Nonetheless, I'd like to see something like that happening, could be a nice little pytest plugin.

NicolasT commented on tests/README.md:
```diff
@@ -0,0 +1,90 @@
+
+#### How to order installation test
+We need ordering between `.1` tests and test that have an installation prerequisite.
+For this we could use the fact that pytest will tend to reduce amount of feature
```
I don't really like this approach, it feels quite brittle. Instead, I'd rather have different sets of tests where some will perform an installation (`installation tests`), and other simply depend on an installation being done before, without expressing this dependency in the test, and without having the test take actions to perform the installation.

NicolasT commented on tests/README.md:
```diff
@@ -0,0 +1,90 @@
+
+
+
+*Proposal*: Handle everything in pytest and handles `.7` in a different test-suite (and other unittesting).
```
See above :)

NicolasT commented on tests/README.md:
```diff
@@ -0,0 +1,90 @@
+*Proposal*: Handle everything in pytest and handles `.7` in a different test-suite (and other unittesting).
+
+*Alternative proposal*: Handle pre-installation in one pytest test-suite.
+    Handle post-installation in another test-suite. A final one for `7.` and different unit-testing
```
Something along those lines yes.

NicolasT commented on tests/README.md:
```diff
@@ -0,0 +1,90 @@
+pytest will need to run ansible. We have two choices here:
+
+
+* subprocess.check_call(‘ansible-playbook’):
```
This is simply the 'setup' phase of 'blackbox' tests, i.e. the tests which don't test 'installation' but post-install functionality only.

NicolasT commented on tests/README.md:
```diff
@@ -0,0 +1,90 @@
+    PRO:
+    - can make assertion about ansible internals
+    CON:
+    - less stable to ansible upgrade
```
Luckily we only support the `make shell` version :wink:

NicolasT commented on tests/README.md:
```diff
@@ -0,0 +1,90 @@
+### Accessing kubernetes services
+
+* telepresence ?
+* multiprocess.run('kubectl port-forward') ?
```
Right now we document service access through `kubectl proxy`, so that's what tests should do.

NicolasT replied:
https://github.com/pytest-dev/pytest-xprocess could be useful here?

NicolasT replied:
Or https://github.com/pytest-dev/pytest-services (not entirely sure though, `xprocess` may be more applicable)

NicolasT reviewed (commented):
Only checked `README` for now, will look into example tests.

NicolasT commented on tests/conftest.py:
```diff
@@ -0,0 +1,12 @@
+import pytest
+
+def pytest_addoption(parser):
```
Could document to export `ANSIBLE_INVENTORY` which removes the need to add custom CLI arguments.

Zempashi replied:
Ah ! but I really would like pytest be aware of the inventory (check if groups are defined), so that it can bail out at the beginning. I'll find a trick (also looking at the environment variable, and prevent passing the `-i` arg in this case...)

NicolasT replied:
Exactly what I was thinking. I'm not so fond of adding arbitrary CLI arguments to existing commands.

NicolasT commented on tests/test_installation.py:
```diff
@@ -0,0 +1,27 @@
+
+@when("I run 'make shell'")
+def run_make_shell():
+    subprocess.check_call('make shell < /dev/null', shell=True)
```
So, we're not really using the `make shell` env. I think this is wrong. Is it possible to launch `make shell` in a subprocess, then keep a reference to it and have e.g. the `I launch the ... playbook` use this subprocess and inject a call to `ansible-playbook ...`?

NicolasT commented on tox.ini:
```diff
@@ -34,3 +34,15 @@ commands =
+commands =
+    pytest tests {posargs:}
+whitelist_externals =
+    make
```
Not invoked in `commands`, shouldn't be needed.

Commit:
Wires new framework in the eve/main.yml

NicolasT commented on eve/main.yml:
```diff
@@ -30,7 +30,23 @@ stages:
         haltOnFailure: true
         command: |-
-          bash ./vendor/bash_unit/bash_unit tests/single-node/test.sh
+          bash tests/single-node/init.sh
```
The current install test comes with its own inventory (which is fine), which declares `/dev/loop0` as storage device, so IMHO it's up to the test to set up this requirement, not some other script ran before the test.

Zempashi replied:
But this one need sudo, and to be on run on the VM. Having something external to the VM, would:
- require to copy git repository over ssh and then run test (which is a performance penalty)
- cannot be done right now by our releng team, and CI environment.

I prefer having a simple contract with tox: provide an inventory, it will run test on it; and then rely on eve orchestration to bootstrap platform and run test on it (sure some test, like failover would never run on single-node test, but could be skipped easily with a `pytest.mark.skipif(condition_on_inventory)`

NicolasT replied:
OK, then let's add a skip if `/dev/loop0` doesn't exist, rather than make tests fail.

Zempashi replied:
`/dev/loop0` is what eve single-node provides, but there is not guarantee for that. I would add some precheck, in the `given` step (something like `playbook/deploy.yml -t assertion`)

NicolasT replied:
Ahaaaa, I may be getting it now.

So basically what you say is: the environment is prepared by Eve (OK!), and the test runs in this environment (OK!), and the test relies on some inventory which is suitable for the test (OK).

In that case, the inventory should be stored in `eve/...`

NicolasT commented on eve/main.yml:
```diff
@@ -30,7 +30,23 @@ stages:
-          bash ./vendor/bash_unit/bash_unit tests/single-node/test.sh
+          bash tests/single-node/init.sh
+    - ShellCommand:
+        name: 'Run the install test suite'
```
Since we'll have several install scenarios (from scratch, upgrade, single-node, multi-node,...) could we already 'implement' this requirement, and run a single-node simple deployment twice?

Zempashi replied:
But we can have "from scratch single-node", "from scratch multi-node". My belief is that all permutation can't be handle in pytest. And then if we decide for a subset of all the possibility, this will prevent a user to come with it's own inventory and run the test-suite. And no user will ever run the test-suite again (Number of platform will consume all the scality.cloud quota required to run test-suite (and the insane time required to run multi-platform tests)), and non-scality member will struggle with integration with cloud provider (which is not nice for open-source project)

NicolasT replied:
OK

NicolasT commented on eve/main.yml:
```diff
@@ -30,7 +30,23 @@ stages:
+        haltOnFailure: true
+        usePTY: true
+        command: |-
+          tox -e test-install -- -s
```
I guess this will be too little.

Zempashi replied:
I don't understand this one. Could you explain a little bit ?

NicolasT replied:
Well, this now runs the `test-install` scenario, but as described above we'll have multiple installation scenarios. It's still unclear to me how this ought to be handled.

NicolasT replied:
Never mind.

NicolasT replied:
What I don't like is how there are many tox targets, one for `test-install`, one for `test`,... I'd rather have one `tox` target, and use `{posargs}` to select relevant tests, then adjust the invocations of `tox` accordingly.

NicolasT commented on requirements.txt:
```diff
@@ -6,8 +6,8 @@
     --hash=sha256:14c9636931309cadabc62298f71758658bc2331cd16a81dce99d359f73e507af
-ansible==2.6.0 \
-    --hash=sha256:3550bf16bf2c34cc88bf3870e91b836e44c1c288d148968c4d855dae2560faf5
+ansible==2.6.1 \
```
We should keep this in sync with the `requirements.txt` used by `make shell` somehow.

Zempashi replied:
So we have ansible here to import it and parse inventory (for testing that prometheus endpoint gather all endpoint it should gather). For installation, this is ansible of `make shell` which is used.
As the internal API of ansible might not be considered as "stable", we might want to keep an "old" version here.
But in the mean time I don't want us to be stuck at an old version of ansible for the run of metal-k8s itself.

NicolasT replied:
Fair enough, indeed everything which invokes Ansible 'like a user' goes through `make shell`, so this is OK.

NicolasT commented on tests/README.md:
```diff
@@ -0,0 +1,40 @@
+There is multiple test-suites for MetalK8s:
```
Skipping this doc for now.

NicolasT commented on tests/conftest.py:
```diff
@@ -0,0 +1,65 @@
+
+from utils.helper import run_ansible_playbook
+
+# logging.basicConfig(level=logging.DEBUG)
```
Remove

Zempashi commented on tests/post_install/test_storage.py:
```diff
@@ -0,0 +1,111 @@
+
+
+@given(parsers.parse(state_storage_string))
+def check_quantity_storage_in_state(quantity, state, kubeconfig):
```
Sadly there is a conflict when I tagged the `then` and the `given` on the same function

NicolasT commented on tests/conftest.py:
```diff
@@ -0,0 +1,65 @@
+
+
+@pytest.fixture
+def pytestbdd_strict_gherkin():
```
Do we really need this?

Zempashi replied:
Yes for the installation test to do `Given-When-Then-When-Then`. You're Stuck at `Given-When-Then` with normal scheme

NicolasT replied:
OK

NicolasT commented on tests/install/features/installation.feature:
```diff
@@ -0,0 +1,10 @@
+Feature: MetalK8s installation
```
I guess we will have various scenarios for this feature (single-node, multi-node,...). How can we make sure all of them run in a separate environment/VM?

NicolasT replied:
Ignore. In fact parametrized over an inventory.

NicolasT commented on tests/install/features/installation.feature:
```diff
@@ -0,0 +1,10 @@
+    When I run 'make shell'
+    And I launch ansible with the 'deploy.yml' playbook
+    Then Playbook should complete without error
+    When I launch ansible with the 'deploy.yml' playbook
```
This is no longer a 'basic installation' :)

NicolasT commented on tests/install/test_installation.py:
```diff
@@ -0,0 +1,32 @@
+
+
+@pytest.fixture
+def pytestbdd_strict_gherkin():
```
Same

NicolasT commented on tests/single-node/init.sh:
```diff
@@ -0,0 +1,13 @@
+echo "Creating loopback block-device and configuring VM"
+sudo truncate -s 20G /kube-lvm || die
```
`die` is not defined

NicolasT reviewed (changes requested):
Partial review, need to follow up.

Zempashi commented on tests/install/make_shell.feature:
```diff
@@ -0,0 +1,6 @@
+Feature: check 'make shell' command
```
Should remove this one until I properly finish it

NicolasT commented on eve/main.yml:
```diff
@@ -29,7 +29,23 @@ stages:
+        command: |-
+          tox -e test-install -- -s
+        env:
+          ANSIBLE_INVENTORY: tests/single-node/inventory/hosts
```
This should point to an Eve directory, not `tests/...`, because the inventory is CI-environment-provisioned.

Zempashi replied:
make sense

NicolasT commented on tests/conftest.py:
```diff
@@ -0,0 +1,64 @@
+
+
+
+def pytest_addoption(parser):
```
Let's stick to one way of passing this on. `ANSIBLE_INVENTORY` in the environment makes most sense (it's an Ansible standard), so let's stick to it and not add code to support another mechanism we don't use anyway?

NicolasT commented on tests/conftest.py:
```diff
@@ -0,0 +1,64 @@
+
+
+@pytest.fixture(scope="session")
+def kubeconfig(inventory):
```
Can we also get this from the environment, `KUBECONFIG`, instead of assuming it's part of the inventory? Requires a bit more setup code before test invocation, but not that much.

NicolasT commented on tests/molecule/requirements.in:
```diff
@@ -0,0 +1,7 @@
+# Can't use `-r../requirements.txt` because that pins package versions, e.g. for
```
I'd rather keep this in `tests/requirements.in`. The corresponding venv can be shared between various tox targets in its configuration.

NicolasT commented on tests/post_install/conftest.py:
```diff
@@ -0,0 +1,42 @@
+            response = requests.get('http://127.0.0.1:8001/')
+        except requests.exceptions.RequestException:
+            return False
+        return bool(response)
```
What's `bool(response)` supposed to be? Alway `True`? In which case is it `False`?

Zempashi replied:
response is true if return code is 200 and False if 404, 500 (But doesn't raise exception)

NicolasT commented on tests/post_install/test_storage.py:
```diff
@@ -0,0 +1,111 @@
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+    kubectl_output = pv_list_process.stdout.read()
```
Should we consider using the Kubernetes API Python bindings instead?

NicolasT commented on tests/single-node/init.sh:
```diff
@@ -0,0 +1,19 @@
+}
+
+echo "Creating loopback block-device and configuring VM"
+sudo truncate -s 256G /kube-lvm || die
```
Would it make sense to require this whole script as root, instead of using `sudo` everywhere?

Also, likely should be part of `eve/...` (cfr. inventory comment earlier)?

NicolasT commented on tests/utils/helper.py:
```diff
@@ -0,0 +1,42 @@
+def run_make_shell(args=None, tmpdir=None, **kwargs):
+    make_args = []
+    if tmpdir:
+        make_args.append('SHELL-ENV={}/.shell-env'.format(tmpdir))
```
Shouldn't this be `SHELL_ENV`?

NicolasT replied:
Also, no real reason to add `/.shell-env` I think?

NicolasT commented on tox.ini:
```diff
@@ -22,6 +22,7 @@ deps =
 commands =
     pip-compile --index --emit-trusted-host --annotate --upgrade --generate-hashes -o "{toxinidir}/requirements.txt" "{toxinidir}/vendor/kubespray/requirements.txt"
     pip-compile --index --emit-trusted-host --annotate --upgrade --generate-hashes -o "{toxinidir}/docs/requirements.txt" "{toxinidir}/docs/requirements.in"
+    pip-compile --index --emit-trusted-host --annotate --upgrade --generate-hashes -o "{toxinidir}/tests/requirements.txt" "{toxinidir}/tests/molecule/requirements.in"
```
Input/output file mismatch. Also, given comment above, not sure this is the right approach.

NicolasT commented on tox.ini:
```diff
@@ -46,8 +47,30 @@ skip_install = true
 whitelist_externals =
     sh
+
+[testenv:test-install]
```
Let's not gratuitously create new envs. Rather re-use some, and use `{posargs}` wisely.

ballot-scality committed:
Get KUBECOFNIG from environment if define

But still fallback on the admin.conf defined in the inventory if not
found.
Fail the test suite if the path in the variable does not exists

ballot-scality committed:
Add KUBECONFIG environment variable in test-install

Commit:
Move inventory into eve

Commit:
Remove inventory cli option

Commit:
Merge molecule requirements

Commit:
Fix make_shell helper

Commit:
Merge tox test command

Commit:
fixup! Remove inventory cli option

Commit:
Move to parsers.re

Commit:
Change scanerio name

Commit:
Revert "Move to parsers.re"

This reverts commit 1fec9ff0076fc8339d34b7a2cd2f07e131ba5c60.

Commit:
Have pytest < 3.6
